native/lib.rs

```diff
@@ -0,0 +1,45 @@
+mod droptimer;
+
+use helium_crypto::{ecc_compact, ed25519, KeyType, Verify};
+use rustler::{Atom, Binary, Error as NifError, ListIterator, NifResult};
+use std::convert::TryFrom;
+
+// -spec verify(
+//     [{Bin :: binary(), [{Signature :: binary(), PubKeyBin :: libp2p_crypto:pubkey_bin()}, ...]}]
+// ) -> ok | {error, Reason :: binary()}.
+#[rustler::nif(name = "verify", schedule = "DirtyCpu")]
+fn verify_1(batch: ListIterator) -> NifResult<Atom> {
+    // let _timer = droptimer::DropTimer::new();
+    for sub_batch in batch.map(|term| term.decode::<(Binary, ListIterator)>()) {
+        let (msg, list) = sub_batch?;
+        for pair in list.map(|term| term.decode::<(Binary, Binary)>()) {
+            let (signature, pubkey_bin) = pair?;
+            let key_type: KeyType = pubkey_bin.first().ok_or(NifError::BadArg).and_then(|&b| {
+                KeyType::try_from(b).map_err(|e| NifError::Term(Box::new(format!("{}", e))))
+            })?;
+            match key_type {
+                KeyType::EccCompact => {
+                    let pk = ecc_compact::PublicKey::try_from(pubkey_bin.as_slice())
+                        .map_err(|e| NifError::Term(Box::new(format!("{}", e))))?;
+                    pk.verify(msg.as_slice(), signature.as_slice())
```

Comment:
You can probably pre-hash the message (signatures are done over hashes) per sub_batch here and pass in a hash to be checked

Reply:
I believe both curves use sha256

Reply:
But that would require using some sort of different/custom verification?